Enable the regex functions to be interrupted

leetcode-1533 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->
Background: 
https://github.com/trinodb/trino/pull/12392 is working on interrupting run away driver threads
This PR is working on making regex functions being able to get interrupted. 

Backport: https://github.com/prestodb/presto/pull/16109. As well as added unit tests
<!-- Answer the following questions to help reviewers and maintainers
understand this PR's scope at a glance:
-->

> Is this change a fix, improvement, new feature, refactoring, or other?
Fix

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
Query Engine

> How would you describe this change to a non-technical end user or system administrator?
Make sure the  regex functions can be interrupted.
Without this change, the regex functions used matcher.search(), which internally used matcher.searchInterruptible() and return -2 when interrupted. 
Some code were doing checking as "offset == -1" for those cases, the -2 flag were ignored and the thread swallowed the interruption. 
For code doing checking as "offset < 0", the change is that instead of returning the slice, it will throw user a trino exception. 

## Related issues, pull requests, and links
Backport: https://github.com/prestodb/presto/pull/16109

<!-- List any issues fixed by this PR, and provide links to other related PRs, upstream release notes, and other useful resources. For example:
* Fixes #issuenumber
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:
Together with https://github.com/trinodb/trino/pull/12392: 

```markdown
# Section
* Fixed driver thread run away issue when processing CPU intensive regular expressions. ({issue}`12392`) ({issue}`12479`)
```

findepi commented:
cc @phd3 @losipiuk @dain @sopel39 @arhimondr @findepi

phd3 commented on core/trino-main/src/test/java/io/trino/sql/gen/TestExpressionCompiler.java:
```diff
@@ -1993,6 +2021,31 @@ public void run()
+        public void run()
+        {
+            try {
+                JoniRegexpFunctions.regexpReplace(utf8Slice(source), joniRegexp(utf8Slice(pattern)));
```
let's add test for regexplike and like for completeness

leetcode-1533 replied:
I added test for regexplike which used the new getMatchingOffset method. JoniRegexpReplaceLambdaFunction used the import getSearchingOffset which had been tested

phd3 commented on core/trino-main/src/test/java/io/trino/sql/gen/TestExpressionCompiler.java:
```diff
@@ -1476,6 +1483,27 @@ public void testFunctionCallRegexp()
+
+        // wait for child thread to get in to terminated state
+        searchChildThread.join();
+        Assert.assertEquals(regularExpressionRunnable.exceptionClause.getErrorCode().getType(), ErrorType.USER_ERROR);
```
interesting, visibility of `exceptionClause` update is always guaranteed here after interruption, so this won't be flaky: https://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.4

> If thread T1 interrupts thread T2, the interrupt by T1 synchronizes-with any point where any other thread (including T2) determines that T2 has been interrupted (by having an InterruptedException thrown or by invoking Thread.interrupted or Thread.isInterrupted).

phd3 replied:
nit: static import for assertequals

arhimondr commented on core/trino-main/src/main/java/io/trino/type/LikeFunctions.java:
```diff
@@ -245,4 +246,14 @@ private static char getEscapeChar(Slice escape)
+            return matcher.matchInterruptible(at, range, Option.NONE);
+        }
+        catch (InterruptedException interruptedException) {
+            throw new TrinoException(GENERIC_USER_ERROR, "Regexp matching interrupted");
```
Recover interrupted flag `Thread.currentThread().interrupt()`

leetcode-1533 replied:
Hey, it is already done in the Joni's library:

  if (interruptCheckCounter++ % INTERRUPT_CHECK_EVERY == 0 && currentThread.isInterrupted()) {
               currentThread.interrupted();
               throw new InterruptedException();
 }

arhimondr replied:
Here is the java doc for the `Thread#interrupted` method:

```
Tests whether the current thread has been interrupted. The interrupted status of the thread is cleared by this method. In other words, if this method were to be called twice in succession, the second call would return false (unless the current thread were interrupted again, after the first call had cleared its interrupted status and before the second call had examined it).
```

The JONI library is compliant with the `InterruptedException` contract. They reset the `interrupted` flag before throwing an exception. Since the `InterruptedException` is being caught the interrupt flag must either be recovered or the thread must be terminated. Since you are simply throwing a different exception the interrupt flag must be recovered to propagate the interrupted status to the upper level code.

leetcode-1533 replied:
I see.. okey, I can recover the thread interrupt flag.

arhimondr commented on core/trino-main/src/main/java/io/trino/operator/scalar/JoniRegexpReplaceLambdaFunction.java:
```diff
@@ -110,8 +110,7 @@ public Slice regexpReplace(
         }
-        while (matcher.search(nextStart, source.length(), Option.DEFAULT) != -1);
-
+        while (JoniRegexpFunctions.getSearchingOffset(matcher, nextStart, source.length()) != -1);
```
nit: static import?

arhimondr commented on core/trino-main/src/main/java/io/trino/operator/scalar/JoniRegexpFunctions.java:
```diff
@@ -408,4 +409,14 @@ public static long regexpCount(@SqlType("varchar(x)") Slice source, @SqlType(Jon
+            return matcher.searchInterruptible(at, range, Option.DEFAULT);
+        }
+        catch (InterruptedException interruptedException) {
+            throw new TrinoException(GENERIC_USER_ERROR, "Regexp matching interrupted");
```
Recover interrupted flag

martint replied:
This should not be "user error". In fact, it should probably not be a TrinoException at all. If this triggers, the query should've already been canceled or aborted and this will be invisible to users.

phd3 replied:
@martint what mechanism would trigger the cancel/abort? theoretically, if `searchInterruptible` running for say 600s,  wouldn't the query still be alive until we kill using https://github.com/trinodb/trino/pull/12392/  (unless some other issues with the query)?

martint replied:
If I remember correctly, at some point there used to be code to interrupt threads when a query is canceled or fails (or maybe we just talked about it). Not sure if that is still there, but if not, then this change won't do anything until such mechanism is reinstated.

leetcode-1533 replied:
Hey, this is the statechange listener's code when remove the task:

```
 private void doRemoveTask(TaskHandle taskHandle)
    {
        List<PrioritizedSplitRunner> splits;
        synchronized (this) {
            tasks.remove(taskHandle);
            splits = taskHandle.destroy();

            // stop tracking splits (especially blocked splits which may never unblock)
            allSplits.removeAll(splits);
            intermediateSplits.removeAll(splits);
            blockedSplits.keySet().removeAll(splits);
            waitingSplits.removeAll(splits);
        }

        // call destroy outside of synchronized block as it is expensive and doesn't need a lock on the task executor
        for (PrioritizedSplitRunner split : splits) {
            split.destroy();
        }

        // record completed stats
        long threadUsageNanos = taskHandle.getScheduledNanos();
        completedTasksPerLevel.incrementAndGet(computeLevel(threadUsageNanos));

        log.debug("Task finished or failed %s", taskHandle.getTaskId());
    }
```

We found it didn't modify the running split, so https://github.com/trinodb/trino/pull/12392 added a periodic job to evaluate each running split to make sure they don't stuck. 

Do you have suggestion as what exception to throw? Maybe GENERIC_INTERNAL_ERROR or EXCEEDED_TIME_LIMIT?

leetcode-1533 replied:
I am thinking lower the timeout threholds to 10s from 10minutes, so user can be notified for this issue. So they can stop sending those complex regular expressions. We should also prevent task level retry from retrying this function.

leetcode-1533 replied:
Recover interrupted flag:
This is already handled by the JONI library after checking the interrupt flag.

arhimondr replied:
> Recover interrupted flag: This is already handled by the JONI library after checking the interrupt flag.

Please check my response above

arhimondr replied:
From what I understand we want to preserve the classification of the given error as a `USER_ERROR`, as :

- For the cases when a Driver is interrupted due to a task failure or query cancellation the type of error doesn't matter
- https://github.com/trinodb/trino/pull/12392 introduces background process that will interrupt drivers that run for more than 10 minutes. In such case the exception thrown will be propagated to the user and must be treated as `USER_ERROR`.

@leetcode-1533 Is my understanding correct?

martint replied:
For the second case, the query should be aborted as a first class action, not as a side effect of tasks being interrupted, so the error type and reason can be indicated at that point.

phd3 replied:
makes sense. So the way I understand is that, as part of periodic interruption in #12392, we should check if the query context still exists, and fail the query with a proper exception there. 

We also know the stacktrace for the thread we're interrupting there, so should be able to provide _some_ information around what exactly ran long.

phd3 commented on core/trino-main/src/test/java/io/trino/sql/gen/TestExpressionCompiler.java:
```diff
@@ -1476,6 +1483,27 @@ public void testFunctionCallRegexp()
+        searchChildThread.start();
+
+        // wait for the child thread to make some progress
+        searchChildThread.join(1000);
```
this seems like reasonable wait to make sure it has indeed started executing. however we could examine thread stacktrace until it's indeed stuck. i'm on the fence if it'd be too defensive.  @findepi @losipiuk wdyt from similar tests?

phd3 commented on core/trino-main/src/main/java/io/trino/type/LikeFunctions.java:
```diff
@@ -245,4 +246,14 @@ private static char getEscapeChar(Slice escape)
+            return matcher.matchInterruptible(at, range, Option.NONE);
+        }
+        catch (InterruptedException interruptedException) {
+            throw new TrinoException(GENERIC_USER_ERROR, "Regexp matching interrupted");
```
same comment about error message

phd3 commented on core/trino-main/src/main/java/io/trino/type/LikeFunctions.java:
```diff
@@ -245,4 +246,14 @@ private static char getEscapeChar(Slice escape)
         throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Escape string must be a single character");
     }
+
+    public static int getMatchingOffset(Matcher matcher, int at, int range)
```
this can be private

phd3 commented on core/trino-main/src/test/java/io/trino/sql/gen/TestExpressionCompiler.java:
```diff
@@ -1476,6 +1483,27 @@ public void testFunctionCallRegexp()
+        // wait for child thread to get in to terminated state
+        searchChildThread.join();
+        Assert.assertEquals(regularExpressionRunnable.exceptionClause.getErrorCode().getType(), ErrorType.USER_ERROR);
+        Assert.assertEquals(regularExpressionRunnable.exceptionClause.getErrorCode().getName(), "GENERIC_USER_ERROR");
```
errorcode implements equals so we can compare at once

phd3 commented on core/trino-main/src/main/java/io/trino/operator/scalar/JoniRegexpFunctions.java:
```diff
@@ -397,7 +398,7 @@ public static long regexpCount(@SqlType("varchar(x)") Slice source, @SqlType(Jon
         while (true) {
             // mather.search returns `source.length` if `nextStart` equals `source.length - 1`.
```
nit: update comment to match new method for future reader

leetcode-1533 commented:
Summary: addressed most of the feedback. Major thing is what exception to throw when interrupted. I am thinking lower the timeout threholds to 10s from 10minutes, so users can be notified for this issue. So they can understand why it get terminated by the system and also stop sending those queries.

leetcode-1533 commented:
If you have better idea as merging LikeVarCharRunnable and RegularExpressionRunnable plz let me know, thanks!

arhimondr commented on core/trino-main/src/test/java/io/trino/sql/TestLikeFunctions.java:
```diff
@@ -254,4 +261,50 @@ public void testUnescapeValidLikePattern()
+            throws Exception
+    {
+        String pattern = "\\((.*,)+(.*\\))";
+        String source = Files.readString(Paths.get(EXTRA_LONG_SOURCE_PATH));
```
`Files.readString` approach works only when the resource is a real file. If it is packaged as a JAR this approach is not going to work. Please use `Resources.toString(Resources.getResource(...), UTF_8)` (here and in other place)

arhimondr replied:
Needs to be addressed

arhimondr commented on core/trino-main/src/test/java/io/trino/sql/gen/TestExpressionCompiler.java:
```diff
@@ -1476,6 +1484,26 @@ public void testFunctionCallRegexp()
     }
 
+    @Test
+    public void testSearchInterruptible()
```
This test looks identical to the test in `TestLikeFunctions`. What's the difference?

Also it doesn't feel like these tests belong in those clases, as technically they don't validate expressions but instead (please correct me if I'm wrong) make direct calls to a specific library (JONI). I would recommend extracting this test into it's own test class.

arhimondr replied:
We should extract it into a separate test class. It doesn't seem like these tests should live in `TestLikeFunctions` / `TestExpressionCompiler`

phd3 commented on core/trino-main/src/main/java/io/trino/operator/scalar/JoniRegexpReplaceLambdaFunction.java:
```diff
@@ -110,8 +110,7 @@ public Slice regexpReplace(
         }
-        while (matcher.search(nextStart, source.length(), Option.DEFAULT) != -1);
-
+        while (JoniRegexpFunctions.getSearchingOffset(matcher, nextStart, source.length()) != -1);
```
@leetcode-1533 is the first comment missed?

tangjiangling commented:
Hey guys, what is the current state of this PR (we have a similar problem in our production environment)? cc @wubiaoi

arhimondr commented on core/trino-main/src/main/java/io/trino/operator/scalar/JoniRegexpFunctions.java:
```diff
@@ -408,4 +409,14 @@ public static long regexpCount(@SqlType("varchar(x)") Slice source, @SqlType(Jon
+            return matcher.searchInterruptible(at, range, Option.DEFAULT);
+        }
+        catch (InterruptedException interruptedException) {
+            throw new TrinoException(GENERIC_USER_ERROR, "Regular expression matching was interrupted, likely because it took too long. Regular expression in the worst case can have a catastrophic amount of backtracking and having exponential time complexity");
```
```suggestion
            Thread.currentThread().interrupt();
            throw new TrinoException(GENERIC_USER_ERROR, "Regular expression matching was interrupted, likely because it took too long. Regular expression in the worst case can have a catastrophic amount of backtracking and having exponential time complexity");
```

arhimondr commented on core/trino-main/src/main/java/io/trino/type/LikeFunctions.java:
```diff
@@ -245,4 +246,14 @@ private static char getEscapeChar(Slice escape)
+            return matcher.matchInterruptible(at, range, Option.NONE);
+        }
+        catch (InterruptedException interruptedException) {
+            throw new TrinoException(GENERIC_USER_ERROR, "Regular expression matching was interrupted, likely because it took too long. Regular expression in the worst case can have a catastrophic amount of backtracking and having exponential time complexity");
```
```suggestion
            Thread.currentThread().interrupt();
            throw new TrinoException(GENERIC_USER_ERROR, "Regular expression matching was interrupted, likely because it took too long. Regular expression in the worst case can have a catastrophic amount of backtracking and having exponential time complexity");
```

arhimondr reviewed (commented):
There are a couple of comments to be addressed. Then we should be able to merge.

arhimondr commented:
Superseded by https://github.com/trinodb/trino/pull/13064
